inverse/beamformer_lcmv.m

```diff
@@ -191,143 +205,189 @@
-        [u, s, v] = svd(real(pinv(lf' * invCy *lf)));
-        eta = u(:,1);
-        lf  = lf * eta;
-        if ~isempty(subspace), lforig = lforig * eta; end
-        dipout.ori{i} = eta;
+          % and compute the leadfield for that orientation
+          lf  = lf * maxpowori;
+          dipout.ori{i} = maxpowori;
+          dipout.eta{i} = eta;
+          if hassubspace, lforig = lforig * maxpowori; end
+      end
     end
-  end
   
-  if hasfilter
-    % use the provided filter
-    filt = dip.filter{i};
-  elseif strcmp(weightnorm,'nai')
-    % Van Veen's Neural Activity Index
-    % below equation is equivalent to following:  
-    % filt = pinv(lf' * invCy * lf) * lf' * invCy; 
-    % filt = filt/sqrt(noise*filt*filt');
-    filt = pinv(sqrt(noise * lf' * invCy_squared * lf)) * lf' *invCy; % based on Sekihara & Nagarajan 2008 eqn. 4.15
-  elseif strcmp(weightnorm,'unitnoisegain')
-    % Unit-noise gain minimum variance (aka Borgiotti-Kaplan) beamformer
-    % below equation is equivalent to following:  
-    % filt = pinv(lf' * invCy * lf) * lf' * invCy; 
-    % filt = filt/sqrt(filt*filt');
-    filt = pinv(sqrtm(lf' * invCy_squared * lf)) * lf' *invCy;     % Sekihara & Nagarajan 2008 eqn. 4.15
-  else
     % construct the spatial filter
-    filt = pinv(lf' * invCy * lf) * lf' * invCy;              % van Veen eqn. 23, use PINV/SVD to cover rank deficient leadfield
+    switch weightnorm
+      case 'nai'
+        % Van Veen's Neural Activity Index
+        % below equation is equivalent to following:  
+        % filt = pinv(lf' * invC * lf) * lf' * invC; 
+        % filt = filt/sqrt(noise*filt*filt'); 
+        % the scaling term in the denominator is sqrt of projected noise, as per eqn. 2.67 of Sekihara & Nagarajan 2008 (S&N)
+        if fixedori
+          filt = pinv(sqrt(noise * lf' * invC_squared * lf)) * lf' *invC; % based on S&N eqn. 4.08
+        else
+          ft_error('vector version of nai weight normalization is not implemented');
+        end
+      case 'unitnoisegain'
+        % filt*filt' = I
+        % Unit-noise gain minimum variance (aka Borgiotti-Kaplan) beamformer
+        % below equation is equivalent to following:  
+        % filt = pinv(lf' * invC * lf) * lf' * invC; 
+        % filt = filt/sqrt(filt*filt');
+        if fixedori
+          filt = pinv(sqrt(lf' * invC_squared * lf)) * lf' *invC; % S&N eqn. 4.15
+        else
+          % compute the matrix that is used for scaling of the filter's rows, as per eqn. 4.83
+          denom = pinv(lf' * invC * lf);
+          gamma = denom * (lf' * invC_squared * lf) * denom;
+          
+          % compute the spatial filter, as per eqn. 4.85
+          filt = diag(1./sqrt(diag(gamma))) * denom * lf' * invC;
+        end
+      case 'arraygain'
+        % filt*lf = ||lf||, applies to scalar leadfield, and to one of the possibilities of the vector version, eqn. 4.75
```

Comment:
What if an already normalized lead field is supplied?

Reply:
if that's the case, or when the on-the-fly computed leadfield has been normalised then it's a bit 'double up' (proper Dutch). Although the lf normalisation in ft_compute_leadfield is per a frobenius norm, yet here it's using another norm...

Reply:
(I looked for a safety net for this, but maybe I overlooked it in the code)

Reply:
Oh, and thanks for assigning yourself as a reviewer, I didn't manage to do so: for some reason github couldn't find you in the reviewer box

Reply:
If a normalized lead field has a flag (`cfg.normalize = 'yes'` or something) it could be caught. I am just thinking that it is maybe not obvious for everyone that normalizing a lead field and an 'array gain beamformer' is the same, so this could be a not so uncommon outcome.

Reply:
He, I think that happened automatically when I left this review! Might not be in the list because I have not contributed that much here (yet ...).

Reply:
Hi, glad to see this getting attention again! Tomorrow is a holiday in Denmark so I'll take a deeper look early next week.